### ISSUE TITLE
Stream reset vs Connection failure

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#21762]
In the event of a QUIC stream reset `SSL_get_error()` will return `SSL_ERROR_SSL`. This is also what it returns in the event of some connection level fatal failure. So, at the moment, it does not seem possible with our APIs to determine the difference between these two things. We need to resolve this problem before alpha.

**Comment by hlandau** [comment]
Reading through the SSL_get_error manpage, it seems like `SSL_ERROR_ZERO_RETURN` corresponds moreorless to a stream reset. But it is also what we return for a FIN.

A new return code from SSL_get_error seems like the cleanest approach but may pose compatibility issues.

An important point of consideration is the distinction between an error code describing why an operation failed, and description of whether a specific condition currently holds on an object. SSL_get_error is more or less oriented towards retrieving information on why the last failed operation failed.

A more compatible option might be to have a simple `SSL_is_read_reset` and `SSL_is_write_reset` predicate APIs which applications can call after receiving `SSL_ERROR_SSL` to determine the reason for the error, or assuming connection failure if none return true.

**Comment by t8m** [comment]
IMO it would be too much error-prone to use `SSL_ERROR_ZERO_RETURN`.

I wonder what compatibility issues we would see if we used a new SSL_ERROR_ value for stream reset.

IMO the applications should be able to handle it with minimal or no changes at all. For single stream apps it really does not matter much if the single stream is reset or the connection is closed. For multi-stream apps they will certainly need new code to handle the streams so they can easily incorporate the new SSL_ERROR value.

**Comment by t8m** [comment]
Hmm... we already have SSL_get_stream_write_error_code() and SSL_get_stream_read_error_code(), isn't it sufficient?

**Comment by hlandau** [comment]
> Hmm... we already have SSL_get_stream_read_state(), SSL_get_stream_write_state(), SSL_get_stream_write_error_code(), and SSL_get_stream_read_error_code(), isn't it sufficient?

Good point, I forgot about this. It looks like we don't really have an issue here.

**Comment by kroeckx** [comment]
SSL_ERROR_ZERO_RETURN is really about normal termination, everything is ok. We really don't want that to mean that an error happened

**Comment by mattcaswell** [comment]
So does this mean that you would have to:

Call `SSL_read(_ex)` and get an error.
Then call `SSL_get_error()` and receive `SSL_ERROR_SSL`
Then call `SSL_get_stream_read_state()` and either get `SSL_STREAM_STATE_RESET_REMOTE` or `SSL_STREAM_STATE_CONN_CLOSED`.

That sounds plausible, but quite long-winded

**Comment by hlandau** [comment]
More or less. You could also just call `SSL_get_stream_read_state` immediately after `SSL_read` fails.
